File: apps/moltnet-cli/embed/agents/retry.py
"""Shared async retry utility for MoltNet eval agents."""

import asyncio
import logging
import os
import random

logger = logging.getLogger(__name__)

_DEFAULT_BASE_DELAY = 2.0
_DEFAULT_MAX_DELAY = 15.0


async def with_retry(
    fn,
    *,
    max_attempts: int | None = None,
    base_delay: float = _DEFAULT_BASE_DELAY,
    max_delay: float = _DEFAULT_MAX_DELAY,
    should_retry=None,
):
    """Retry an async callable with exponential backoff + jitter.

    Args:
        fn: Async callable that takes no arguments. Use a lambda or
            functools.partial to bind arguments before passing.
        max_attempts: Maximum number of attempts. Defaults to AGENT_MAX_RETRIES
            env var (default 3).
        base_delay: Base delay in seconds for first retry (default 2.0).
        max_delay: Maximum delay cap in seconds (default 15.0).
        should_retry: Optional callable ``(exc) -> bool``. Return False to abort
            immediately without retrying. Defaults to retrying on any exception.

    Returns:
        The return value of ``fn()`` on success.

    Raises:
        The last exception raised by ``fn()`` after all attempts are exhausted,
        or immediately if ``should_retry`` returns False.
    """
    if max_attempts is None:
        max_attempts = int(os.environ.get("AGENT_MAX_RETRIES", "3"))

    last_exc = None
    for attempt in range(max_attempts):
        try:
            return await fn()
        except Exception as exc:  # noqa: BLE001
            last_exc = exc
            if should_retry is not None and not should_retry(exc):
                raise
            if attempt < max_attempts - 1:
                delay = min(base_delay * (2**attempt), max_delay)
                jitter = random.uniform(0, 0.5 * delay)
                wait = delay + jitter
                logger.warning(
                    "[retry] attempt %d/%d failed: %s. Retrying in %.1fs...",
                    attempt + 1,
                    max_attempts,
                    exc,
                    wait,
                )
                await asyncio.sleep(wait)

    logger.error("[retry] all %d attempts failed", max_attempts)
    raise last_exc

File: apps/moltnet-cli/embed/judge/retry.js
/* eslint-disable no-console */

/**
 * Retry an async function with exponential backoff + jitter.
 *
 * @param {() => Promise<T>} fn - Async callable to retry (no arguments)
 * @param {object} [opts]
 * @param {number} [opts.maxAttempts] - Max attempts (default: JUDGE_MAX_RETRIES env var or 3)
 * @param {number} [opts.baseDelayMs] - Base delay in ms (default: 2000)
 * @param {number} [opts.maxDelayMs] - Max delay cap in ms (default: 15000)
 * @param {(err: Error) => boolean} [opts.shouldRetry] - Return false to abort immediately (default: always retry)
 * @returns {Promise<T>}
 * @template T
 */
export async function withRetry(fn, opts = {}) {
  const maxAttempts = opts.maxAttempts ?? parseInt(process.env.JUDGE_MAX_RETRIES ?? '3', 10);
  const baseDelayMs = opts.baseDelayMs ?? 2000;
  const maxDelayMs = opts.maxDelayMs ?? 15000;
  const shouldRetry = opts.shouldRetry ?? (() => true);

  let lastErr;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (!shouldRetry(err)) {
        throw err;
      }
      if (attempt < maxAttempts - 1) {
        const delay = Math.min(baseDelayMs * 2 ** attempt, maxDelayMs);
        const jitter = Math.random() * 0.5 * delay;
        const wait = Math.round(delay + jitter);
        console.error(
          `[retry] attempt ${attempt + 1}/${maxAttempts} failed: ${err.message ?? err}. Retrying in ${wait}ms...`,
        );
        await new Promise((resolve) => setTimeout(resolve, wait));
      }
    }
  }
  console.error(`[retry] all ${maxAttempts} attempts failed`);
  throw lastErr;
}
